flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/committer/FilesCommitter.java

```diff
@@ -0,0 +1,314 @@
+        WriteResult result = commitRequest.getCommittable().committable();
+        RowDelta rowDelta = table.newRowDelta().scanManifestsWith(workerPool);
+        Arrays.stream(result.dataFiles()).forEach(rowDelta::addRows);
+        Arrays.stream(result.deleteFiles()).forEach(rowDelta::addDeletes);
+
+        String flinkJobId = commitRequest.getCommittable().jobID();
+        long checkpointId = commitRequest.getCommittable().checkpointId();
+        commitOperation(
+            commitRequest,
+            rowDelta,
+            dataFilesNum,
+            deleteFilesNum,
+            "rowDelta",
+            flinkJobId,
+            checkpointId);
+      }
+    }
+  }
+
+  private void commitOperation(
+      CommitRequest<FilesCommittable> commitRequest,
+      SnapshotUpdate<?> operation,
+      int numDataFiles,
+      int numDeleteFiles,
+      String description,
+      String newFlinkJobId,
+      long checkpointId) {
+    commitOperation(
+        Lists.newArrayList(commitRequest),
+        operation,
+        numDataFiles,
+        numDeleteFiles,
+        description,
+        newFlinkJobId,
+        checkpointId);
+  }
+
+  private void commitOperation(
+      Collection<CommitRequest<FilesCommittable>> commitRequests,
+      SnapshotUpdate<?> operation,
+      int numDataFiles,
+      int numDeleteFiles,
+      String description,
+      String newFlinkJobId,
+      long checkpointId) {
+    try {
+      commitOperation(
+          operation, numDataFiles, numDeleteFiles, description, newFlinkJobId, checkpointId);
+    } catch (AlreadyExistsException | UnsupportedOperationException e) {
+      commitRequests.forEach(request -> request.signalFailedWithKnownReason(e));
+    } catch (Exception e) {
+      commitRequests.forEach(request -> request.signalFailedWithUnknownReason(e));
+    }
+  }
+
+  private void commitOperation(
+      SnapshotUpdate<?> operation,
+      int numDataFiles,
+      int numDeleteFiles,
+      String description,
+      String newFlinkJobId,
+      long checkpointId) {
+    LOG.info(
+        "Committing {} with {} data files and {} delete files to table {}",
+        description,
+        numDataFiles,
+        numDeleteFiles,
+        table);
+    // custom snapshot metadata properties will be overridden if they conflict with internal ones
+    // used by the sink.
+    snapshotProperties.forEach(operation::set);
+    operation.set(MAX_COMMITTED_CHECKPOINT_ID, Long.toString(checkpointId));
+    operation.set(FLINK_JOB_ID, newFlinkJobId);
+
+    long start = System.currentTimeMillis();
+    operation.commit(); // abort is automatically called if this fails.
+    long duration = System.currentTimeMillis() - start;
+    LOG.info("Committed in {} ms", duration);
+  }
+
+  static long getMaxCommittedCheckpointId(Table table, String flinkJobId) {
+    Snapshot snapshot = table.currentSnapshot();
+    long lastCommittedCheckpointId = INITIAL_CHECKPOINT_ID;
```

Comment:
Is the table.history ordered by the time?
Do we commit anything without `MAX_COMMITTED_CHECKPOINT_ID` with the given `FLINK_JOB_ID`?

We might be able to use the things above to get a smaller, more readable method.

Also we usually avoid `get`. Either use `find`, or just use `maxCommittedCheckpointId`

```
private static long maxCommittedCheckpointId(Table table, String flinkJobId) {
  for (HistoryEntry history : table.history()) {
     Map<String, String> summary = table.snapshot(history.snapshotId()).summary();
     if (flinkJobId.equals(summary..get(FLINK_JOB_ID))) {
        return Long.parseLong(summary..get(MAX_COMMITTED_CHECKPOINT_ID));
     }
  }

  return INITIAL_CHECKPOINT_ID;
}
```
Or:
```
  private static long maxCommittedCheckpointId(org.apache.iceberg.Table table, String flinkJobId) {
    return table.history().stream()
            .map(history -> table.snapshot(history.snapshotId()).summary())
            .filter(summary -> summary.containsKey(FLINK_JOB_ID))
            .findFirst()
            .map(summary -> Long.parseLong(summary.get(MAX_COMMITTED_CHECKPOINT_ID)))
            .orElse(INITIAL_CHECKPOINT_ID);
  }
```

Reply:
https://github.com/apache/iceberg/blob/c7ce6bbafe476391807694c65f242c53a334ba0c/flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/committer/FilesCommitter.java#L266-L273